Source/Tests/LinuxSyscalls/Syscalls.cpp

```diff
@@ -226,89 +211,133 @@ static bool IsSupportedByInterpreter(std::string const &Filename) {
-    auto Path = FEX::HLE::_SyscallHandler->FM.GetEmulatedPath(pathname, true);
-    if (!Path.empty() && std::filesystem::exists(Path, ec)) {
-      Filename = Path;
+  // AT_EMPTY_PATH is only used if the pathname is empty.
+  const bool IsFDExec = (Args.flags & AT_EMPTY_PATH) && strlen(pathname) == 0;
+  std::string FDExecEnv;
+
+  if (IsFDExec) {
+    Type = ELFLoader::ELFContainer::GetELFType(Args.dirfd);
+
+    if (!IsShebangFD(Args.dirfd) && Type == ELFLoader::ELFContainer::ELFType::TYPE_NONE) {
+      // If our interpeter doesn't support this file format AND ELF format is NONE then ENOEXEC
+      // binfmt_misc could end up handling this case but we can't know that without parsing binfmt_misc ourselves
+      // Return -ENOEXEC until proven otherwise
+      return -ENOEXEC;
+    }
+  }
+  else
+  {
+    // Check the rootfs if it is available first
+    if (pathname[0] == '/') {
+      auto Path = FEX::HLE::_SyscallHandler->FM.GetEmulatedPath(pathname, true);
+      if (!Path.empty() && std::filesystem::exists(Path, ec)) {
+        Filename = Path;
+      }
+      else {
+        Filename = pathname;
+      }
     }
     else {
       Filename = pathname;
     }
-  }
-  else {
-    Filename = pathname;
-  }
 
-  bool exists = std::filesystem::exists(Filename, ec);
-  if (ec || !exists) {
-    return -ENOENT;
-  }
+    bool exists = std::filesystem::exists(Filename, ec);
+    if (ec || !exists) {
+      return -ENOENT;
+    }
+
+    int pid = getpid();
+
+    char PidSelfPath[50];
+    snprintf(PidSelfPath, 50, "/proc/%i/exe", pid);
 
-  int pid = getpid();
+    if (strcmp(pathname, "/proc/self/exe") == 0 ||
+        strcmp(pathname, "/proc/thread-self/exe") == 0 ||
+        strcmp(pathname, PidSelfPath) == 0) {
+      // If the application is trying to execve `/proc/self/exe` or its variants,
+      // then we need to redirect this path to the true application path.
+      // This is because this path is a symlink to the executing application, which is always `FEXInterpreter` or `FEXLoader`.
+      // ex: JRE and shapez.io do this self-execution.
+      Filename = FEX::HLE::_SyscallHandler->Filename();
+    }
 
-  char PidSelfPath[50];
-  snprintf(PidSelfPath, 50, "/proc/%i/exe", pid);
+    Type = ELFLoader::ELFContainer::GetELFType(Filename);
 
-  if (strcmp(pathname, "/proc/self/exe") == 0 ||
-      strcmp(pathname, "/proc/thread-self/exe") == 0 ||
-      strcmp(pathname, PidSelfPath) == 0) {
-    // If pointing to self then redirect to the application
-    // JRE and shapez.io does this
-    Filename = FEX::HLE::_SyscallHandler->Filename();
+    if (!IsShebangFilename(Filename) && Type == ELFLoader::ELFContainer::ELFType::TYPE_NONE) {
+      // If our interpeter doesn't support this file format AND ELF format is NONE then ENOEXEC
+      // binfmt_misc could end up handling this case but we can't know that without parsing binfmt_misc ourselves
+      // Return -ENOEXEC until proven otherwise
+      return -ENOEXEC;
+    }
```

Comment:
This is repeated twice now. Instead, we can write the result of IsShebangFilename to a boolean in each branch and use a single `!is_shebang && Type == ...` check after the branch.

Reply:
Duplication removed.